### PR TITLE
TF inference optimization tutorial

### PR DESCRIPTION
**Pull request by louie-tsai**
## Description

This tutorial will guide you how to optimize a pre-trained model for a better inference performance, and also analyze the model pb files before and after the inference optimizations

## External Dependencies
pandas, matplotlib

## Checklist
Administrative
- [x] Review sample design with the appropriate [Domain Expert](https://github.com/oneapi-src/oneAPI-samples/wiki/Reviewers-and-Domain-Experts): Anand, Gandhimathi ; Ramesh, AG; Wang, Wei V
- [x] Sample has dependencies and/or include Third Party binaries. If yes inform the oneAPI Code Samples Project Manager: @JoeOster

Code Development
- [x] Implement coding guidelines and ensure code quality. [see wiki for details](https://github.com/oneapi-src/oneAPI-samples/wiki/General-Code-Guidelines)
- [x] Adhere to readme template 
- [ ] Enforce format via clang-format config file
- [x] Adhere to sample.json specification. https://github.com/oneapi-src/oneAPI-samples/wiki/sample-json-specification
- [x] Ensure/create CI test configurations for sample (ciTests field) https://github.com/oneapi-src/oneAPI-samples/wiki/sample-json-ci-test-object
- [ ] Run jsonlint on sample.json to verify json syntax. www.jsonlint.com

Security and Legal
- [x] OSPDT Approval (see @JoeOster for assistance)
- [ ] Compile using the following compiler flags and fix any warnings, the falgs are: "-Wall -Wformat-security -Werror=format-security"
- [x] Protex (see @JoeOster for assistance)
- [x] IPLDT Scan (see @JoeOster for assistance) ran on 05/18/2021
- [ ] Virus scan
- [x] Code is written in Python. if checked, Bandit scans are required- ) Posted to teams on 05/20/2021

Review
- [ ] Review DPC++ code with Paul Peterseon. (GitHub User: pmpeter1)
- [ ] Review readme with Tom Lenth(@tomlenth) and/or Joe Oster(@JoeOster
- [ ] Tested using Dev Cloud when applicable

### CODE REVIEW

**Comment by louie-tsai** [comment]
attached a pdf with results from previous run.
[tutorial_optimize_TensorFlow_pretrained_model_0426.pdf](https://github.com/oneapi-src/oneAPI-samples/files/6476070/tutorial_optimize_TensorFlow_pretrained_model_0426.pdf)

**Comment by JoeOster** [comment]
@louie-tsai I cleaned up and modified the PR Description

**Comment by JoeOster** [comment]
@louie-tsai Which domain expert did you use to review this with? Put it in the first line for the Admin checklist

**Review by tomlenth — COMMENTED** [review]
The Readme is not following the guideline template.   For a similar AI sample, see Distributed TensorFlow with Horovod.  All sections need to be included.  In addition, the purpose really needs to be expanded to better promote the sample.  it should be a good 3-4 sentence explanation for the importance.  You may want to explain the TF ops differences up front. The key implementation details should include what technology is used, or what's different about this sample than other AI samples.

File: AI-and-Analytics/Features-and-Functionality/IntelTensorFlow_InferenceOptimization/sample.json [hunk 1]
```diff
@@ -0,0 +1,20 @@
+{
+	"guid": "9d32f194-8667-41d3-865d-d43e9983c471",
+	"name": "Tutorial : Optimize TensorFlow pre-trained model for inference",
```
Comment (JoeOster) [inline AI-and-Analytics/Features-and-Functionality/IntelTensorFlow_InferenceOptimization/sample.json:3]: Why are we adding a tutorial to the name, usually this is for organizational purposes, if that's the case are you planning on modifying the other samples?

File: AI-and-Analytics/Features-and-Functionality/IntelTensorFlow_InferenceOptimization/sample.json [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+	"categories": ["Toolkit/oneAPI AI And Analytics/Features And Functionality"],
+	"description": "This tutorial will guide you how to optimize a pre-trained model for a better inference performance, and also analyze the model pb files before and after the inference optimizations.",
+	"builder": ["cli"],
+	"languages": [{
+		"python": {}
```
Comment (JoeOster) [inline AI-and-Analytics/Features-and-Functionality/IntelTensorFlow_InferenceOptimization/sample.json:8]: As this is python, I will need a bandit scan from you prior to merging

**Review by JoeOster — CHANGES_REQUESTED** [review]

**Review by ganand1 — DISMISSED** [review]

File: AI-and-Analytics/Features-and-Functionality/IntelTensorFlow_InferenceOptimization/sample.json (hunk 2 above)
Comment (louie-tsai) [reply AI-and-Analytics/Features-and-Functionality/IntelTensorFlow_InferenceOptimization/sample.json:8]: **just got below warning. 
any suggestion for possible fix or replacement of os.system?**

Run started:2021-05-17 23:03:03.590150

Test results:
>> Issue: [B605:start_process_with_a_shell] Starting a process with a shell, possible injection detected, security issue.
   Severity: High   Confidence: High
   Location: IntelTensorFlow_InferenceOptimization/scripts/profile_utils.py:21
20      def do_command(command):
21          os.system(command)
22

--------------------------------------------------

Code scanned:
        Total lines of code: 1421
        Total lines skipped (#nosec): 0

Run metrics:
        Total issues (by severity):
                Undefined: 0
                Low: 0
                Medium: 0
                High: 1
        Total issues (by confidence):
                Undefined: 0
                Low: 0
                Medium: 0
                High: 1
Files skipped (0):

Comment (JoeOster) [reply AI-and-Analytics/Features-and-Functionality/IntelTensorFlow_InferenceOptimization/sample.json:8]: take a look at: https://docs.openstack.org/bandit/1.4.0/plugins/start_process_with_a_shell.html

File: AI-and-Analytics/Features-and-Functionality/IntelTensorFlow_InferenceOptimization/README.md [hunk 3]
```diff
@@ -0,0 +1,69 @@
+# Tutorial : Optimize TensorFlow pre-trained model for inference
+This tutorial will guide you how to optimize a pre-trained model for a better inference performance, and also 
+analyze the model pb files before and after the inference optimizations.  
+
+| Optimized for                     | Description
+|:---                               |:---
+| OS                                | Linux* Ubuntu* 18.04
```
Comment (JoeOster) [inline AI-and-Analytics/Features-and-Functionality/IntelTensorFlow_InferenceOptimization/README.md:7]: remove Linux, just show ubuntu

File: AI-and-Analytics/Features-and-Functionality/IntelTensorFlow_InferenceOptimization/README.md [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+Third party program Licenses can be found here: [third-party-programs.txt](https://github.com/oneapi-src/oneAPI-samples/blob/master/third-party-programs.txt)
+
+## Build and Run the Sample
+
+### Pre-requirement
+
+> NOTE: No action is required if users use Intel DevCloud as their environment. 
+  Please refer to [Intel oneAPI DevCloud](https://intelsoftwaresites.secure.force.com/devcloud/oneapi) for Intel DevCloud.
+
+ 1. **Intel AI Analytics Toolkit**
```
Comment (JoeOster) [inline AI-and-Analytics/Features-and-Functionality/IntelTensorFlow_InferenceOptimization/README.md:38]: Please add the "®" symbol after Intel

File: AI-and-Analytics/Features-and-Functionality/IntelTensorFlow_InferenceOptimization/README.md [hunk 5]
```diff
@@ -0,0 +1,69 @@
+# Tutorial : Optimize TensorFlow pre-trained model for inference
+This tutorial will guide you how to optimize a pre-trained model for a better inference performance, and also 
+analyze the model pb files before and after the inference optimizations.  
+
+| Optimized for                     | Description
+|:---                               |:---
+| OS                                | Linux* Ubuntu* 18.04 
+| Hardware                          | Intel® Xeon® Scalable processor family or newer
+| Software                          | Intel® oneAPI AI Analytics Toolkit
```
Comment (JoeOster) [inline AI-and-Analytics/Features-and-Functionality/IntelTensorFlow_InferenceOptimization/README.md:9]: I think it would be a good idea to put a link here to the AI toolkit

**Review by JoeOster — CHANGES_REQUESTED** [review]

File: AI-and-Analytics/Features-and-Functionality/IntelTensorFlow_InferenceOptimization/sample.json (hunk 2 above)
Comment (louie-tsai) [reply AI-and-Analytics/Features-and-Functionality/IntelTensorFlow_InferenceOptimization/sample.json:8]: Remove os.system for bandit

new result : 
ality$ bandit IntelTensorFlow_InferenceOptimization/
[main]  INFO    profile include tests: None
[main]  INFO    profile exclude tests: None
[main]  INFO    cli include tests: None
[main]  INFO    cli exclude tests: None
[main]  INFO    running on Python 2.7.17
[manager]       WARNING Skipping directory (IntelTensorFlow_InferenceOptimization/), use -r flag to scan contents
Run started:2021-05-19 23:54:31.366297

Test results:
        No issues identified.

Code scanned:
        Total lines of code: 0
        Total lines skipped (#nosec): 0

Run metrics:
        Total issues (by severity):
                Undefined: 0
                Low: 0
                Medium: 0
                High: 0
        Total issues (by confidence):
                Undefined: 0
                Low: 0
                Medium: 0
                High: 0
Files skipped (0):

**Comment by louie-tsai** [comment]
> @louie-tsai Which domain expert did you use to review this with? Put it in the first line for the Admin checklist

added them accordingly

File: AI-and-Analytics/Features-and-Functionality/IntelTensorFlow_InferenceOptimization/sample.json (hunk 1 above)
Comment (louie-tsai) [reply AI-and-Analytics/Features-and-Functionality/IntelTensorFlow_InferenceOptimization/sample.json]: changed it accordingly.

File: AI-and-Analytics/Features-and-Functionality/IntelTensorFlow_InferenceOptimization/README.md (hunk 3 above)
Comment (louie-tsai) [reply AI-and-Analytics/Features-and-Functionality/IntelTensorFlow_InferenceOptimization/README.md]: removed it accordingly.

File: AI-and-Analytics/Features-and-Functionality/IntelTensorFlow_InferenceOptimization/README.md (hunk 4 above)
Comment (louie-tsai) [reply AI-and-Analytics/Features-and-Functionality/IntelTensorFlow_InferenceOptimization/README.md]: added it accordingly.

File: AI-and-Analytics/Features-and-Functionality/IntelTensorFlow_InferenceOptimization/README.md (hunk 5 above)
Comment (louie-tsai) [reply AI-and-Analytics/Features-and-Functionality/IntelTensorFlow_InferenceOptimization/README.md]: put the link accordingly.

**Comment by louie-tsai** [comment]
@JoeOster  
addressed the feedback accordingly. please help me to merge it.

**Review by JoeOster — DISMISSED** [review]
Looks good to me, @pmpeter1 and @tomlenth I need both of your approvals prior to merging

**Comment by louie-tsai** [comment]
> The Readme is not following the guideline template. For a similar AI sample, see Distributed TensorFlow with Horovod. All sections need to be included. In addition, the purpose really needs to be expanded to better promote the sample. it should be a good 3-4 sentence explanation for the importance. You may want to explain the TF ops differences up front. The key implementation details should include what technology is used, or what's different about this sample than other AI samples.

Hi Tom,

I tried to address your comments in README.
The difference among other sample should be addressed in purpose section. no other sample has the same purpose.
The technology for sure is Intel TensorFlow. no need to bring it up again.

**Comment by louie-tsai** [comment]
@tomlenth 
Could you share your feedback? thanks!

**Review by JoeOster — APPROVED** [review]
approving for ci only

**Comment by louie-tsai** [comment]
@JoeOster 

The CI issue is " No module named 'tensorflow' " which should be environment issue.

I got successful run on my environment. 
Capture

overall, I will use conda run -n tensorflow instead to force the ci use tensorflow env for this run.

please run CI again to test the new change in sample.json.

**Review by JoeOster — DISMISSED** [review]
approved for ci only

File: AI-and-Analytics/Features-and-Functionality/IntelTensorFlow_InferenceOptimization/README.md [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+Those optimizations includes:  
+* Converting variables to constants.
+* Removing training-only operations like checkpoint saving.
+* Stripping out parts of the graph that are never reached.
+* Removing debug operations like CheckNumerics.
+* Folding batch normalization ops into the pre-calculated weights.
+* Fusing common operations into unified versions.
+
+## Key implementation details
+This tutorial contains one Jupyter notebook and couple python scripts listed below.
```
Comment (tomlenth) [inline AI-and-Analytics/Features-and-Functionality/IntelTensorFlow_InferenceOptimization/README.md:24]: this is grammatically incorrect.  Please replace "couple" with "three".

**Review by tomlenth — CHANGES_REQUESTED** [review]
One minor text change required.

File: AI-and-Analytics/Features-and-Functionality/IntelTensorFlow_InferenceOptimization/README.md (hunk 6 above)
Comment (louie-tsai) [reply AI-and-Analytics/Features-and-Functionality/IntelTensorFlow_InferenceOptimization/README.md]: thanks! addressed it accordingly.

**Comment by louie-tsai** [comment]
@JoeOster  
correct sample.json and also fixed some grammatic problems.
please help me to validate the ci testing.

**Review by JoeOster — APPROVED** [review]
approving for ci

**Review by JoeOster — COMMENTED** [review]
approving for ci only

File: AI-and-Analytics/Features-and-Functionality/IntelTensorFlow_InferenceOptimization/README.md [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+|:---                               |:---
+| OS                                | Ubuntu* 18.04 
+| Hardware                          | Intel® Xeon® Scalable processor family or newer
+| Software                          | [Intel® oneAPI AI Analytics Toolkit](https://software.intel.com/content/www/us/en/develop/tools/oneapi/ai-analytics-toolkit.html)
+| What you will learn               | Optimize a pre-trained model for a better inference performance
+| Time to complete                  | 30 minutes
+
+## Purpose
+Show users the importance of inference optimization on performance, and also analyze TensorFlow ops difference in pre-trained models before/after the optimizations.  
+Those optimizations includes:
```
Comment (fitchbe) [inline AI-and-Analytics/Features-and-Functionality/IntelTensorFlow_InferenceOptimization/README.md:15]: Those optimizations **includes** --> Those optimizations **include**

**Review by fitchbe — APPROVED** [review]

**Review by JoeOster — DISMISSED** [review]
approving for ci

**Comment by louie-tsai** [comment]
@JoeOster 

I have no idea why we don't have tensorflow in the CI environment.
This error seems to be the environment issue.
any idea for possible cause?

With the latest change, we could get a report.html from CI run.
please try it again and then get the report.html file for further analysis.

File: AI-and-Analytics/Features-and-Functionality/IntelTensorFlow_InferenceOptimization/README.md (hunk 7 above)
Comment (louie-tsai) [reply AI-and-Analytics/Features-and-Functionality/IntelTensorFlow_InferenceOptimization/README.md]: addressed it accordingly.

**Review by JoeOster — DISMISSED** [review]
approved for ci only

**Review by JoeOster — APPROVED** [review]
approving for ci

**Review by JoeOster — APPROVED** [review]

**Review by JoeOster — APPROVED** [review]

**Review by JoeOster — DISMISSED** [review]

**Review by JoeOster — APPROVED** [review]

**Review by JoeOster — APPROVED** [review]
